bulk_indexer.go

```diff
@@ -162,17 +158,23 @@ type bulkIndexerItem struct {
 	Index      string
 	Action     string
 	DocumentID string
-	Body       io.Reader
+	Body       io.WriterTo
 }
 
 // add encodes an item in the buffer.
 func (b *bulkIndexer) add(item bulkIndexerItem) error {
 	b.writeMeta(item.Index, item.Action, item.DocumentID)
-	if _, err := io.CopyBuffer(b.writer, item.Body, b.copybuf[:]); err != nil {
-		return err
+	if _, err := item.Body.WriteTo(&b.buf); err != nil {
+		return fmt.Errorf("failed to write bulk indexer item: %w", err)
+	}
+	if _, err := b.buf.Write([]byte("\n")); err != nil {
+		return fmt.Errorf("failed to write newline: %w", err)
 	}
-	if _, err := b.writer.Write([]byte("\n")); err != nil {
-		return err
+	if b.gzipw != nil && b.buf.Len() > 512*1024 {
```

Comment:
I would prefer if this is configured in the `newBulkIndexer`. This will allow the compress threshold to be a function of `Config.FlushBytes` (could be half or 3/4 of it). We could call it `compressThreshold`.